File: pipeline/scan.py
#!/usr/bin/env python3
"""
Deterministic scanner for ~/dev projects.

Collects raw git info, language indicators, file flags, TODO/FIXME counts.
Outputs JSON to stdout. Accepts DEV_ROOT and EXCLUDE_DIRS as arguments.

Usage:
    python3 scan.py <dev_root> <exclude_csv>
"""

import hashlib
import json
import os
import subprocess
import sys
from datetime import datetime, timezone
from pathlib import Path

LANGUAGE_INDICATORS: dict[str, str] = {
    "package.json": "JavaScript/TypeScript",
    "pyproject.toml": "Python",
    "setup.py": "Python",
    "requirements.txt": "Python",
    "Cargo.toml": "Rust",
    "go.mod": "Go",
    "Gemfile": "Ruby",
    "build.gradle": "Java/Kotlin",
    "pom.xml": "Java",
    "mix.exs": "Elixir",
    "Package.swift": "Swift",
    "composer.json": "PHP",
}

SOURCE_EXTENSIONS: set[str] = {
    ".py", ".ts", ".tsx", ".js", ".jsx", ".rs", ".go", ".rb", ".java", ".kt",
    ".ex", ".exs", ".swift", ".php", ".c", ".cpp", ".h",
}

SKIP_WALK_DIRS: set[str] = {
    "node_modules", ".venv", ".git", "__pycache__", "dist", "build",
    ".next", "target", ".tox", "venv", "env",
}


def path_hash(absolute_path: str) -> str:
    """Stable identity hash from absolute path."""
    return hashlib.sha256(absolute_path.encode()).hexdigest()[:16]


def run_git(cwd: str, *args: str) -> str | None:
    try:
        result = subprocess.run(
            ["git", *args],
            cwd=cwd,
            capture_output=True,
            text=True,
            timeout=5,
        )
        return result.stdout.strip() if result.returncode == 0 else None
    except (subprocess.TimeoutExpired, FileNotFoundError):
        return None


def get_git_info(path: str) -> dict:
    if not (Path(path) / ".git").exists():
        return {
            "isRepo": False,
            "lastCommitDate": None,
            "lastCommitMessage": None,
            "branch": None,
            "remoteUrl": None,
            "commitCount": 0,
            "daysInactive": None,
            "isDirty": False,
            "untrackedCount": 0,
            "modifiedCount": 0,
            "stagedCount": 0,
            "ahead": 0,
            "behind": 0,
            "recentCommits": [],
            "branchCount": 0,
            "stashCount": 0,
        }

    last_date = run_git(path, "log", "-1", "--format=%aI")
    last_msg = run_git(path, "log", "-1", "--format=%s")
    branch = run_git(path, "rev-parse", "--abbrev-ref", "HEAD")
    remote = run_git(path, "remote", "get-url", "origin")
    count_str = run_git(path, "rev-list", "--count", "HEAD")
    commit_count = int(count_str) if count_str else 0

    days_inactive = None
    if last_date:
        try:
            last_dt = datetime.fromisoformat(last_date)
            days_inactive = (datetime.now(timezone.utc) - last_dt).days
        except ValueError:
            pass

    # Working tree status
    status_output = run_git(path, "status", "--porcelain")
    is_dirty = False
    untracked = 0
    modified = 0
    staged = 0
    if status_output:
        is_dirty = True
        for line in status_output.splitlines():
            if len(line) < 2:
                continue
            x, y = line[0], line[1]
            if x == "?" and y == "?":
                untracked += 1
            if y == "M" or x == "M":
                modified += 1
            if x in ("A", "M", "R", "D") and y != "?":
                staged += 1

    # Ahead/behind remote
    ahead_count = 0
    behind_count = 0
    if branch:
        ab_output = run_git(path, "rev-list", "--count", "--left-right", f"@{{upstream}}...HEAD")
        if ab_output:
            parts = ab_output.split()
            if len(parts) == 2:
                try:
                    behind_count = int(parts[0])
                    ahead_count = int(parts[1])
                except ValueError:
                    pass

    # Recent commits
    recent_commits: list[dict] = []
    log_output = run_git(path, "log", "-10", "--format=%H|%aI|%s")
    if log_output:
        for line in log_output.splitlines():
            parts = line.split("|", 2)
            if len(parts) == 3:
                recent_commits.append({
                    "hash": parts[0],
                    "date": parts[1],
                    "message": parts[2],
                })

    # Branch count
    branch_output = run_git(path, "branch", "--list")
    branch_count = len(branch_output.splitlines()) if branch_output else 0

    # Stash count
    stash_output = run_git(path, "stash", "list")
    stash_count = len(stash_output.splitlines()) if stash_output else 0

    return {
        "isRepo": True,
        "lastCommitDate": last_date,
        "lastCommitMessage": last_msg,
        "branch": branch,
        "remoteUrl": remote,
        "commitCount": commit_count,
        "daysInactive": days_inactive,
        "isDirty": is_dirty,
        "untrackedCount": untracked,
        "modifiedCount": modified,
        "stagedCount": staged,
        "ahead": ahead_count,
        "behind": behind_count,
        "recentCommits": recent_commits,
        "branchCount": branch_count,
        "stashCount": stash_count,
    }


def detect_languages(path: str) -> dict:
    detected: list[str] = []
    primary = None

    for indicator, lang in LANGUAGE_INDICATORS.items():
        if (Path(path) / indicator).exists():
            if lang not in detected:
                detected.append(lang)
            if primary is None:
                primary = lang

    if (Path(path) / "tsconfig.json").exists():
        if "JavaScript/TypeScript" in detected:
            primary = "TypeScript"
        elif "TypeScript" not in detected:
            detected.append("TypeScript")
            if primary is None:
                primary = "TypeScript"

    return {"primary": primary, "detected": detected}


def check_files(path: str) -> dict:
    p = Path(path)
    has_tests = any(
        (p / d).exists()
        for d in ["tests", "test", "__tests__", "spec", "src/tests", "src/__tests__"]
    ) or bool(list(p.glob("*test*")))

    return {
        "readme": (p / "README.md").exists() or (p / "readme.md").exists(),
        "tests": has_tests,
        "env": (p / ".env").exists(),
        "envExample": (p / ".env.example").exists(),
        "dockerfile": (p / "Dockerfile").exists(),
        "dockerCompose": (
            (p / "docker-compose.yml").exists()
            or (p / "docker-compose.yaml").exists()
            or (p / "compose.yml").exists()
        ),
    }


def check_cicd(path: str) -> dict:
    p = Path(path)
    return {
        "githubActions": (p / ".github" / "workflows").exists(),
        "circleci": (p / ".circleci").exists(),
        "travis": (p / ".travis.yml").exists(),
        "gitlabCi": (p / ".gitlab-ci.yml").exists(),
    }


def check_deployment(path: str) -> dict:
    p = Path(path)
    return {
        "fly": (p / "fly.toml").exists(),
        "vercel": (p / "vercel.json").exists(),
        "netlify": (p / "netlify.toml").exists(),
    }


def count_todos(path: str) -> tuple[int, int, int]:
    """Walk source files, counting TODOs, FIXMEs, and total lines of code."""
    todo_count = 0
    fixme_count = 0
    loc_count = 0

    for root, dirs, files in os.walk(path):
        dirs[:] = [d for d in dirs if d not in SKIP_WALK_DIRS]
        for fname in files:
            if Path(fname).suffix not in SOURCE_EXTENSIONS:
                continue
            fpath = os.path.join(root, fname)
            try:
                with open(fpath, "r", errors="ignore") as f:
                    for line in f:
                        loc_count += 1
                        if "TODO" in line:
                            todo_count += 1
                        if "FIXME" in line:
                            fixme_count += 1
            except (PermissionError, OSError):
                continue

    return todo_count, fixme_count, loc_count


def get_description(path: str) -> str | None:
    pkg = Path(path) / "package.json"
    if pkg.exists():
        try:
            data = json.loads(pkg.read_text())
            desc = data.get("description")
            if desc:
                return desc
        except (json.JSONDecodeError, OSError):
            pass

    pyproject = Path(path) / "pyproject.toml"
    if pyproject.exists():
        try:
            text = pyproject.read_text()
            for line in text.splitlines():
                if line.strip().startswith("description"):
                    parts = line.split("=", 1)
                    if len(parts) == 2:
                        return parts[1].strip().strip('"').strip("'")
        except OSError:
            pass

    return None


FRAMEWORK_MAP_JS: dict[str, str] = {
    "next": "nextjs",
    "react": "react",
    "vue": "vue",
    "@angular/core": "angular",
    "express": "express",
    "fastify": "fastify",
    "svelte": "svelte",
    "nuxt": "nuxt",
    "@remix-run/react": "remix",
    "gatsby": "gatsby",
}

FRAMEWORK_MAP_RUST: dict[str, str] = {
    "axum": "axum",
    "actix-web": "actix",
    "rocket": "rocket",
    "warp": "warp",
}

FRAMEWORK_MAP_PYTHON: list[tuple[str, str]] = [
    ("fastapi", "fastapi"),
    ("django", "django"),
    ("flask", "flask"),
    ("starlette", "starlette"),
]

SERVICE_DEPS: dict[str, str] = {
    "@supabase/supabase-js": "supabase",
    "posthog-js": "posthog",
    "posthog-node": "posthog",
    "stripe": "stripe",
    "firebase": "firebase",
    "firebase-admin": "firebase",
    "@aws-sdk": "aws",
    "@prisma/client": "prisma",
    "mongoose": "mongodb",
    "@sentry": "sentry",
}

ENV_KEY_PREFIXES: list[tuple[str, str]] = [
    ("SUPABASE_", "supabase"),
    ("POSTHOG_", "posthog"),
    ("NEXT_PUBLIC_POSTHOG", "posthog"),
    ("STRIPE_", "stripe"),
    ("FIREBASE_", "firebase"),
    ("AWS_", "aws"),
    ("DATABASE_URL", "database"),
    ("SENTRY_", "sentry"),
    ("OPENAI_", "openai"),
    ("ANTHROPIC_", "anthropic"),
]

LOCKFILE_MAP: list[tuple[str, str]] = [
    ("pnpm-lock.yaml", "pnpm"),
    ("package-lock.json", "npm"),
    ("yarn.lock", "yarn"),
    ("bun.lockb", "bun"),
    ("Cargo.lock", "cargo"),
    ("uv.lock", "uv"),
    ("poetry.lock", "poetry"),
    ("Pipfile.lock", "pipenv"),
]


def _read_json(path: str) -> dict | None:
    """Read a JSON file, returning None on any error."""
    try:
        return json.loads(Path(path).read_text())
    except (json.JSONDecodeError, OSError):
        return None


def detect_framework(path: str) -> str | None:
    """Detect the primary framework from dependency files."""
    p = Path(path)

    # Check package.json
    pkg = _read_json(str(p / "package.json"))
    if pkg:
        all_deps: dict = {}
        all_deps.update(pkg.get("dependencies", {}) or {})
        all_deps.update(pkg.get("devDependencies", {}) or {})
        for dep_name, framework in FRAMEWORK_MAP_JS.items():
            if dep_name in all_deps:
                return framework

    # Check Cargo.toml
    cargo = p / "Cargo.toml"
    if cargo.exists():
        try:
            text = cargo.read_text()
            for dep_name, framework in FRAMEWORK_MAP_RUST.items():
                # Match both [dependencies] table entries and inline
                if dep_name in text:
                    return framework
        except OSError:
            pass

    # Check pyproject.toml and requirements.txt
    pyproject = p / "pyproject.toml"
    requirements = p / "requirements.txt"
    py_text = ""
    if pyproject.exists():
        try:
            py_text += pyproject.read_text()
        except OSError:
            pass
    if requirements.exists():
        try:
            py_text += requirements.read_text()
        except OSError:
            pass
    if py_text:
        for dep_name, framework in FRAMEWORK_MAP_PYTHON:
            if dep_name in py_text:
                return framework

    return None


def detect_scripts(path: str) -> list[str]:
    """Extract script names from package.json."""
    pkg = _read_json(str(Path(path) / "package.json"))
    if pkg:
        scripts = pkg.get("scripts")
        if isinstance(scripts, dict):
            return list(scripts.keys())
    return []


def detect_services(path: str) -> list[str]:
    """Detect external services from deps and .env key prefixes."""
    p = Path(path)
    services: set[str] = set()

    # Check package.json dependencies
    pkg = _read_json(str(p / "package.json"))
    if pkg:
        all_deps: dict = {}
        all_deps.update(pkg.get("dependencies", {}) or {})
        all_deps.update(pkg.get("devDependencies", {}) or {})
        for dep_name in all_deps:
            for pattern, service in SERVICE_DEPS.items():
                if dep_name == pattern or dep_name.startswith(pattern + "/"):
                    services.add(service)

    # Check .env key prefixes (KEYS ONLY, never values)
    for env_file in [".env", ".env.local", ".env.development"]:
        env_path = p / env_file
        if env_path.exists():
            try:
                with open(str(env_path), "r", errors="ignore") as f:
                    for line in f:
                        line = line.strip()
                        if not line or line.startswith("#"):
                            continue
                        key = line.split("=", 1)[0].strip()
                        for prefix, service in ENV_KEY_PREFIXES:
                            if key.startswith(prefix):
                                services.add(service)
            except (PermissionError, OSError):
                continue

    return sorted(services)


def detect_package_manager(path: str) -> str | None:
    """Detect package manager from lockfiles."""
    p = Path(path)
    for filename, manager in LOCKFILE_MAP:
        if (p / filename).exists():
            return manager
    return None


def detect_license(path: str) -> bool:
    """Check for LICENSE or LICENSE.md."""
    p = Path(path)
    return (p / "LICENSE").exists() or (p / "LICENSE.md").exists()


def has_language_indicators(path: str) -> bool:
    """Check if a directory contains any language indicator files."""
    return any((Path(path) / f).exists() for f in LANGUAGE_INDICATORS)


def scan_project(abs_path: str) -> dict:
    name = os.path.basename(abs_path)
    git_info = get_git_info(abs_path)
    languages = detect_languages(abs_path)
    files = check_files(abs_path)
    cicd = check_cicd(abs_path)
    deployment = check_deployment(abs_path)
    todo_count, fixme_count, loc_estimate = count_todos(abs_path)
    description = get_description(abs_path)
    framework = detect_framework(abs_path)
    scripts = detect_scripts(abs_path)
    services = detect_services(abs_path)
    package_manager = detect_package_manager(abs_path)
    license_found = detect_license(abs_path)

    return {
        "name": name,
        "path": abs_path,
        "pathHash": path_hash(abs_path),
        **git_info,
        "languages": languages,
        "files": files,
        "cicd": cicd,
        "deployment": deployment,
        "todoCount": todo_count,
        "fixmeCount": fixme_count,
        "description": description,
        "framework": framework,
        "scripts": scripts,
        "services": services,
        "locEstimate": loc_estimate,
        "packageManager": package_manager,
        "license": license_found,
    }


def main() -> None:
    if len(sys.argv) < 3:
        print("Usage: scan.py <dev_root> <exclude_csv>", file=sys.stderr)
        sys.exit(1)

    dev_root = os.path.expanduser(sys.argv[1])
    exclude_dirs = set(d.strip() for d in sys.argv[2].split(",") if d.strip())

    if not os.path.isdir(dev_root):
        print(json.dumps({"error": f"{dev_root} not found"}))
        sys.exit(1)

    projects = []
    for entry in sorted(Path(dev_root).iterdir()):
        if not entry.is_dir():
            continue
        if entry.name.startswith("."):
            continue
        if entry.name in exclude_dirs:
            continue
        # Skip non-project folders unless they have language indicators or are git repos
        abs_path = str(entry)
        if not (entry / ".git").exists() and not has_language_indicators(abs_path):
            continue
        projects.append(scan_project(abs_path))

    output = {
        "scannedAt": datetime.now(timezone.utc).isoformat(),
        "projectCount": len(projects),
        "projects": projects,
    }

    print(json.dumps(output, indent=2))


if __name__ == "__main__":
    main()

File: src/lib/pipeline-native/scan.ts
/**
 * TypeScript-native scan — deterministic scanner for ~/dev projects.
 *
 * Port of pipeline/scan.py. Collects raw git info, language indicators,
 * file flags, TODO/FIXME counts. Returns structured data (no subprocess).
 */

import { createHash } from "crypto";
import { execFileSync } from "child_process";
import fs from "fs";
import path from "path";

// ---------------------------------------------------------------------------
// Constants (mirroring scan.py)
// ---------------------------------------------------------------------------

const LANGUAGE_INDICATORS: Record<string, string> = {
  "package.json": "JavaScript/TypeScript",
  "pyproject.toml": "Python",
  "setup.py": "Python",
  "requirements.txt": "Python",
  "Cargo.toml": "Rust",
  "go.mod": "Go",
  "Gemfile": "Ruby",
  "build.gradle": "Java/Kotlin",
  "pom.xml": "Java",
  "mix.exs": "Elixir",
  "Package.swift": "Swift",
  "composer.json": "PHP",
  "index.html": "HTML/CSS",
};

const SOURCE_EXTENSIONS = new Set([
  ".py", ".ts", ".tsx", ".js", ".jsx", ".rs", ".go", ".rb", ".java", ".kt",
  ".ex", ".exs", ".swift", ".php", ".c", ".cpp", ".h",
]);

const SKIP_WALK_DIRS = new Set([
  "node_modules", ".venv", ".git", "__pycache__", "dist", "build",
  ".next", "target", ".tox", "venv", "env",
]);

const SERVICE_DEPS: Record<string, string> = {
  "@supabase/supabase-js": "supabase",
  "posthog-js": "posthog",
  "posthog-node": "posthog",
  stripe: "stripe",
  firebase: "firebase",
  "firebase-admin": "firebase",
  "@aws-sdk": "aws",
  "@prisma/client": "prisma",
  mongoose: "mongodb",
  "@sentry": "sentry",
};

const ENV_KEY_PREFIXES: [string, string][] = [
  ["SUPABASE_", "supabase"],
  ["POSTHOG_", "posthog"],
  ["NEXT_PUBLIC_POSTHOG", "posthog"],
  ["STRIPE_", "stripe"],
  ["FIREBASE_", "firebase"],
  ["AWS_", "aws"],
  ["DATABASE_URL", "database"],
  ["SENTRY_", "sentry"],
  ["OPENAI_", "openai"],
  ["ANTHROPIC_", "anthropic"],
];

const LOCKFILE_MAP: [string, string][] = [
  ["pnpm-lock.yaml", "pnpm"],
  ["package-lock.json", "npm"],
  ["yarn.lock", "yarn"],
  ["bun.lockb", "bun"],
  ["Cargo.lock", "cargo"],
  ["uv.lock", "uv"],
  ["poetry.lock", "poetry"],
  ["Pipfile.lock", "pipenv"],
];

const LINTER_FILES = [
  ".eslintrc", ".eslintrc.js", ".eslintrc.json", ".eslintrc.yml",
  "eslint.config.js", "eslint.config.mjs", "eslint.config.ts",
  ".prettierrc", ".prettierrc.js", ".prettierrc.json",
  "biome.json", "biome.jsonc",
  ".flake8", ".pylintrc", "pyproject.toml",
  ".rubocop.yml", "rustfmt.toml",
];

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

export function pathHash(absolutePath: string): string {
  return createHash("sha256").update(absolutePath).digest("hex").slice(0, 16);
}

function runGit(cwd: string, ...args: string[]): string | null {
  try {
    return execFileSync("git", args, {
      cwd,
      timeout: 5_000,
      encoding: "utf-8",
      stdio: ["pipe", "pipe", "pipe"],
    }).trim() || null;
  } catch {
    return null;
  }
}

function existsAt(dir: string, file: string): boolean {
  return fs.existsSync(path.join(dir, file));
}

function readJsonSafe(filePath: string): Record<string, unknown> | null {
  try {
    return JSON.parse(fs.readFileSync(filePath, "utf-8"));
  } catch {
    return null;
  }
}

// ---------------------------------------------------------------------------
// Detection functions
// ---------------------------------------------------------------------------

function getGitInfo(projectPath: string): Record<string, unknown> {
  if (!fs.existsSync(path.join(projectPath, ".git"))) {
    return {
      isRepo: false,
      lastCommitDate: null,
      lastCommitMessage: null,
      branch: null,
      remoteUrl: null,
      commitCount: 0,
      daysInactive: null,
      isDirty: false,
      untrackedCount: 0,
      modifiedCount: 0,
      stagedCount: 0,
      ahead: 0,
      behind: 0,
      recentCommits: [],
      branchCount: 0,
      stashCount: 0,
    };
  }

  const lastDate = runGit(projectPath, "log", "-1", "--format=%aI");
  const lastMsg = runGit(projectPath, "log", "-1", "--format=%s");
  const branch = runGit(projectPath, "rev-parse", "--abbrev-ref", "HEAD");
  const remote = runGit(projectPath, "remote", "get-url", "origin");
  const countStr = runGit(projectPath, "rev-list", "--count", "HEAD");
  const commitCount = countStr ? parseInt(countStr, 10) : 0;

  let daysInactive: number | null = null;
  if (lastDate) {
    try {
      const lastDt = new Date(lastDate);
      daysInactive = Math.floor((Date.now() - lastDt.getTime()) / (1000 * 60 * 60 * 24));
    } catch {
      // ignore
    }
  }

  // Working tree status
  const statusOutput = runGit(projectPath, "status", "--porcelain");
  let isDirty = false;
  let untracked = 0;
  let modified = 0;
  let staged = 0;
  if (statusOutput) {
    isDirty = true;
    for (const line of statusOutput.split("\n")) {
      if (line.length < 2) continue;
      const x = line[0];
      const y = line[1];
      if (x === "?" && y === "?") untracked++;
      if (y === "M" || x === "M") modified++;
      if ("AMRD".includes(x) && y !== "?") staged++;
    }
  }

  // Ahead/behind
  let aheadCount = 0;
  let behindCount = 0;
  if (branch) {
    const abOutput = runGit(projectPath, "rev-list", "--count", "--left-right", "@{upstream}...HEAD");
    if (abOutput) {
      const parts = abOutput.split(/\s+/);
      if (parts.length === 2) {
        behindCount = parseInt(parts[0], 10) || 0;
        aheadCount = parseInt(parts[1], 10) || 0;
      }
    }
  }

  // Recent commits
  const recentCommits: Array<{ hash: string; date: string; message: string }> = [];
  const logOutput = runGit(projectPath, "log", "-10", "--format=%H|%aI|%s");
  if (logOutput) {
    for (const line of logOutput.split("\n")) {
      const parts = line.split("|", 3);
      if (parts.length === 3) {
        recentCommits.push({ hash: parts[0], date: parts[1], message: parts[2] });
      }
    }
  }

  // Branch count
  const branchOutput = runGit(projectPath, "branch", "--list");
  const branchCount = branchOutput ? branchOutput.split("\n").length : 0;

  // Stash count
  const stashOutput = runGit(projectPath, "stash", "list");
  const stashCount = stashOutput ? stashOutput.split("\n").length : 0;

  return {
    isRepo: true,
    lastCommitDate: lastDate,
    lastCommitMessage: lastMsg,
    branch,
    remoteUrl: remote,
    commitCount,
    daysInactive,
    isDirty,
    untrackedCount: untracked,
    modifiedCount: modified,
    stagedCount: staged,
    ahead: aheadCount,
    behind: behindCount,
    recentCommits,
    branchCount,
    stashCount,
  };
}

function detectLanguages(projectPath: string): { primary: string | null; detected: string[] } {
  const detected: string[] = [];
  let primary: string | null = null;

  for (const [indicator, lang] of Object.entries(LANGUAGE_INDICATORS)) {
    if (existsAt(projectPath, indicator)) {
      if (!detected.includes(lang)) detected.push(lang);
      if (primary === null) primary = lang;
    }
  }

  if (existsAt(projectPath, "tsconfig.json")) {
    if (detected.includes("JavaScript/TypeScript")) {
      primary = "TypeScript";
    } else if (!detected.includes("TypeScript")) {
      detected.push("TypeScript");
      if (primary === null) primary = "TypeScript";
    }
  }

  return { primary, detected };
}

function checkFiles(projectPath: string): Record<string, boolean> {
  const p = projectPath;
  const testDirs = ["tests", "test", "__tests__", "spec", "src/tests", "src/__tests__"];
  const hasTests = testDirs.some((d) => fs.existsSync(path.join(p, d))) ||
    (() => {
      try {
        return fs.readdirSync(p).some((f) => f.includes("test"));
      } catch {
        return false;
      }
    })();

  const hasLinter = LINTER_FILES.some((f) => existsAt(p, f));
  const hasLockfile = LOCKFILE_MAP.some(([f]) => existsAt(p, f));

  return {
    readme: existsAt(p, "README.md") || existsAt(p, "readme.md"),
    tests: hasTests,
    env: existsAt(p, ".env"),
    envExample: existsAt(p, ".env.example"),
    dockerfile: existsAt(p, "Dockerfile"),
    dockerCompose: existsAt(p, "docker-compose.yml") || existsAt(p, "docker-compose.yaml") || existsAt(p, "compose.yml"),
    linterConfig: hasLinter,
    license: detectLicense(p),
    lockfile: hasLockfile,
  };
}

function checkCicd(projectPath: string): Record<string, boolean> {
  return {
    githubActions: fs.existsSync(path.join(projectPath, ".github", "workflows")),
    circleci: fs.existsSync(path.join(projectPath, ".circleci")),
    travis: existsAt(projectPath, ".travis.yml"),
    gitlabCi: existsAt(projectPath, ".gitlab-ci.yml"),
  };
}

function checkDeployment(projectPath: string): Record<string, boolean> {
  return {
    fly: existsAt(projectPath, "fly.toml"),
    vercel: existsAt(projectPath, "vercel.json"),
    netlify: existsAt(projectPath, "netlify.toml"),
  };
}

function countTodos(projectPath: string): [number, number, number] {
  let todoCount = 0;
  let fixmeCount = 0;
  let locCount = 0;

  function walk(dir: string) {
    let entries: fs.Dirent[];
    try {
      entries = fs.readdirSync(dir, { withFileTypes: true });
    } catch {
      return;
    }

    for (const entry of entries) {
      if (entry.isDirectory()) {
        if (!SKIP_WALK_DIRS.has(entry.name)) {
          walk(path.join(dir, entry.name));
        }
      } else if (entry.isFile()) {
        const ext = path.extname(entry.name);
        if (!SOURCE_EXTENSIONS.has(ext)) continue;
        try {
          const content = fs.readFileSync(path.join(dir, entry.name), "utf-8");
          // Match Python iterator semantics: split on \n but don't count
          // a trailing empty element from a final newline
          const lines = content.split("\n");
          const lineCount = lines.length > 0 && lines[lines.length - 1] === "" ? lines.length - 1 : lines.length;
          for (let li = 0; li < lineCount; li++) {
            locCount++;
            if (lines[li].includes("TODO")) todoCount++;
            if (lines[li].includes("FIXME")) fixmeCount++;
          }
        } catch {
          // permission error, skip
        }
      }
    }
  }

  walk(projectPath);
  return [todoCount, fixmeCount, locCount];
}

function getDescription(projectPath: string): string | null {
  const pkg = readJsonSafe(path.join(projectPath, "package.json"));
  if (pkg) {
    const desc = pkg.description;
    if (typeof desc === "string" && desc) return desc;
  }

  const pyprojectPath = path.join(projectPath, "pyproject.toml");
  if (fs.existsSync(pyprojectPath)) {
    try {
      const text = fs.readFileSync(pyprojectPath, "utf-8");
      for (const line of text.split("\n")) {
        if (line.trim().startsWith("description")) {
          const parts = line.split("=", 2);
          if (parts.length === 2) {
            return parts[1].trim().replace(/^["']|["']$/g, "");
          }
        }
      }
    } catch {
      // ignore
    }
  }

  return null;
}

function detectScripts(projectPath: string): string[] {
  const pkg = readJsonSafe(path.join(projectPath, "package.json"));
  if (pkg) {
    const scripts = pkg.scripts;
    if (scripts && typeof scripts === "object") {
      return Object.keys(scripts);
    }
  }
  return [];
}

function detectServices(projectPath: string): string[] {
  const services = new Set<string>();

  // Check package.json
  const pkg = readJsonSafe(path.join(projectPath, "package.json"));
  if (pkg) {
    const allDeps: Record<string, unknown> = {
      ...((pkg.dependencies as Record<string, unknown>) ?? {}),
      ...((pkg.devDependencies as Record<string, unknown>) ?? {}),
    };
    for (const depName of Object.keys(allDeps)) {
      for (const [pattern, service] of Object.entries(SERVICE_DEPS)) {
        if (depName === pattern || depName.startsWith(pattern + "/")) {
          services.add(service);
        }
      }
    }
  }

  // Check .env key prefixes
  for (const envFile of [".env", ".env.local", ".env.development"]) {
    const envPath = path.join(projectPath, envFile);
    if (!fs.existsSync(envPath)) continue;
    try {
      const content = fs.readFileSync(envPath, "utf-8");
      for (const line of content.split("\n")) {
        const trimmed = line.trim();
        if (!trimmed || trimmed.startsWith("#")) continue;
        const key = trimmed.split("=", 1)[0].trim();
        for (const [prefix, service] of ENV_KEY_PREFIXES) {
          if (key.startsWith(prefix)) services.add(service);
        }
      }
    } catch {
      // ignore
    }
  }

  return [...services].sort();
}

function detectPackageManager(projectPath: string): string | null {
  for (const [filename, manager] of LOCKFILE_MAP) {
    if (existsAt(projectPath, filename)) return manager;
  }
  return null;
}

function detectLicense(projectPath: string): boolean {
  return existsAt(projectPath, "LICENSE") || existsAt(projectPath, "LICENSE.md");
}

function hasLanguageIndicators(projectPath: string): boolean {
  return Object.keys(LANGUAGE_INDICATORS).some((f) => existsAt(projectPath, f));
}

// ---------------------------------------------------------------------------
// Lightweight directory enumeration (no deep scan)
// ---------------------------------------------------------------------------

export interface ProjectDir {
  name: string;
  absPath: string;
  pathHash: string;
}

export function listProjectDirs(devRoot: string, excludeDirs: string[], includeNonGitDirs = true): ProjectDir[] {
  if (!fs.existsSync(devRoot) || !fs.statSync(devRoot).isDirectory()) {
    throw new Error(`Scan root not found: ${devRoot}`);
  }

  const excludeSet = new Set(excludeDirs.filter(Boolean));
  const dirs: ProjectDir[] = [];
  const entries = fs.readdirSync(devRoot, { withFileTypes: true });

  entries.sort((a, b) => a.name.localeCompare(b.name));

  for (const entry of entries) {
    if (!entry.isDirectory()) continue;
    if (entry.name.startsWith(".")) continue;
    if (excludeSet.has(entry.name)) continue;

    const absPath = path.join(devRoot, entry.name);
    if (!includeNonGitDirs && !fs.existsSync(path.join(absPath, ".git")) && !hasLanguageIndicators(absPath)) {
      continue;
    }

    dirs.push({ name: entry.name, absPath, pathHash: pathHash(absPath) });
  }

  return dirs;
}

// ---------------------------------------------------------------------------
// Main scan function
// ---------------------------------------------------------------------------

export interface ScanOutput {
  scannedAt: string;
  projectCount: number;
  projects: Array<Record<string, unknown>>;
}

export function scanProject(absPath: string): Record<string, unknown> {
  const name = path.basename(absPath);
  const gitInfo = getGitInfo(absPath);
  const languages = detectLanguages(absPath);
  const files = checkFiles(absPath);
  const cicd = checkCicd(absPath);
  const deployment = checkDeployment(absPath);
  const [todoCount, fixmeCount, locEstimate] = countTodos(absPath);
  const description = getDescription(absPath);
  const framework = null; // Phase 61W: framework detection moved to LLM enrichment

  let liveUrl: string | null = null;
  const pkg = readJsonSafe(path.join(absPath, "package.json"));
  if (pkg) {
    const homepage = pkg.homepage;
    if (typeof homepage === "string" && homepage.trim()) {
      liveUrl = homepage.trim();
    }
  }

  const scripts = detectScripts(absPath);
  const services = detectServices(absPath);
  const packageManager = detectPackageManager(absPath);
  const license = detectLicense(absPath);

  return {
    name,
    path: absPath,
    pathHash: pathHash(absPath),
    ...gitInfo,
    languages,
    files,
    cicd,
    deployment,
    todoCount,
    fixmeCount,
    description,
    framework,
    liveUrl,
    scripts,
    services,
    locEstimate,
    packageManager,
    license,
  };
}

export function scanAll(devRoot: string, excludeDirs: string[], includeNonGitDirs = true): ScanOutput {
  if (!fs.existsSync(devRoot) || !fs.statSync(devRoot).isDirectory()) {
    throw new Error(`Scan root not found: ${devRoot}`);
  }

  const excludeSet = new Set(excludeDirs.filter(Boolean));
  const projects: Array<Record<string, unknown>> = [];
  const entries = fs.readdirSync(devRoot, { withFileTypes: true });

  // Sort for deterministic order
  entries.sort((a, b) => a.name.localeCompare(b.name));

  for (const entry of entries) {
    if (!entry.isDirectory()) continue;
    if (entry.name.startsWith(".")) continue;
    if (excludeSet.has(entry.name)) continue;

    const absPath = path.join(devRoot, entry.name);
    if (!includeNonGitDirs && !fs.existsSync(path.join(absPath, ".git")) && !hasLanguageIndicators(absPath)) {
      continue;
    }

    projects.push(scanProject(absPath));
  }

  return {
    scannedAt: new Date().toISOString(),
    projectCount: projects.length,
    projects,
  };
}
